File: comm/interfaces_src/ServerCommunication.java
package interfaces;

import java.io.*;
import java.net.*;
import java.util.*;

/**
 * <code>ServerCommunication</code> class used to access interprocess communication server
 * Programmer: Christian Wagner
 * Date Created: 3/11/2020
 * Date Modified: 11/02/2020
 * Version: 2.0
 */
//
public class ServerCommunication {
    // debug messages flag
    private final boolean DEBUG_FLAG = false;

    // defaults
    private final String DEFAULT_HOST = "localhost"; // default host name
    private final int DEFAULT_PORT = 8189; // default port number
    private final int COORDINATION_PORT = 18000; // port for communication coordination
    private final int RETRY_CONNECTION_WAIT = 1; // time to wait, in seconds, between retries
    private final int RETRY_CONNECTION_COUNT = 10; // number of times to retry before giving up

    // object variables
    private String host; // host name
    private int port; // port number
    private int PID; // process identification number
    private Socket socket; // client socket
    private DataOutputStream write; // client socket write stream
    private BufferedReader read; // client socket read stream

    /**
     * Default constructor for <code>ServerCommunication</code> class
     */
    public ServerCommunication() {
        this.host = DEFAULT_HOST;
        this.port = DEFAULT_PORT;
        this.PID = 0;
    }

    /**
     * Overloaded constructor for <code>ServerCommunication</code> class
     * @param host host name
     * @param port port number being served on
     * @param PID process identification number
     */
    public ServerCommunication(String host, int port, int PID) {
        this.host = host;
        this.port = port;
        this.PID = PID;
    }

    /**
     * Getter for the process identification number
     * @return process identification number
     */
    public int getPID() {
        return this.PID;
    }


    /**
     * Connect to server using given host and port parameters, only tries once
     * @return true if connection was established, false otherwise
     */
    private boolean connect() {
        try {
            // do coordination
            socket = new Socket(host, COORDINATION_PORT);
            write = new DataOutputStream(socket.getOutputStream());
            read = new BufferedReader(new InputStreamReader(socket.getInputStream()));

            // ensure connection was made, if not, try again
            if(!socket.isConnected()) {
                if(DEBUG_FLAG) System.out.println("Connection on coordination port failed!");

                // clear vars
                socket = null;
                write = null;
                read = null;

                // try to make connection again
                for(int i = 0; i < RETRY_CONNECTION_COUNT; i++) {
                    if(DEBUG_FLAG) System.out.println("Retrying connection on coordination port");

                    socket = new Socket(host, COORDINATION_PORT);
                    write = new DataOutputStream(socket.getOutputStream());
                    read = new BufferedReader(new InputStreamReader(socket.getInputStream()));

                    try {
                        Thread.sleep(RETRY_CONNECTION_WAIT * 1000); // wait
                    } catch(InterruptedException e){
                        return false;
                    }

                    if(socket.isConnected()) break; // connection was made

                    // clear and restart
                    socket = null;
                    write = null;
                    read = null;
                }

                if(!socket.isConnected()) return false; // connection failed
            }

            if(DEBUG_FLAG) System.out.println("Connected to coordination port!");

            String output = port + "\n";
            write.write(output.getBytes()); // communicate desired port
            String status = read.readLine();
            if(status.equals("a")) {
                this.disconnect(); // don't care about return
                // connection accepted, port is open
                socket = new Socket(host, port);
                write = new DataOutputStream(socket.getOutputStream());
                read = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                this.write.writeUTF("PID=" + this.PID + "\n"); // send PID information
                return true;
            } else {
                this.disconnect(); // don't care about return

                if(DEBUG_FLAG) System.out.println("Coordination failed.");
                if(DEBUG_FLAG) System.out.println("Selected port is being used.");

                return false;
            }
        } catch (IOException e) { return false; }
    }

    /**
     * Disconnect from the server
     * @return true if disconnect was successful, false otherwise
     */
    private boolean disconnect() {
        try {
            this.write.writeUTF("exit\n"); // signal exit
            write.close(); // close open write stream
            read.close(); // close open read stream
            socket.close(); // close open connection
            write = null; // clear variable
            read = null; // clear variable
            socket = null; // clear variable
        } catch (IOException e) { return false; }

        return true;
    }

    /**
     * Open the current connection to the server
     * @return trye if open was successful, false otherwise
     */
    public boolean open() {
        if(socket != null) if(socket.isConnected()) return true; // socket was already open
        return this.connect();
    }

    /**
     * Close the current connection to the server
     * @return true if close was successful, false otherwise
     */
    public boolean close() {
        if(socket != null) if(socket.isClosed()) return true; // socket was already closed
        if(socket == null) return true; // socket was already closed
        else return this.disconnect(); // trying to disconnect
    }

    /**
     * Send data to the server for the queue
     * @param msg the data to send
     * @return returns true if successful, false otherwise
     */
    public boolean WriteQueue(IgluMessage msg) {
        if(!socket.isConnected()) return false; // catch socket not being connected

        try {
            // send data to server
            this.write.writeUTF("QW\n");
            this.write.writeUTF((msg == null) ? "NULL" : msg.getTxData()); // send from IgluMessage object
            this.write.writeUTF("\n");
        } catch (IOException e) { return false; }

        return true;
    }

    /**
     * Check for items in the queue
     * @return returns true if items are in the queue, false otherwise
     */
    private boolean CheckQueue() {
        if(!socket.isConnected()) return false; // catch socket not being connected

        try {
            this.write.writeUTF("QC\n");
            String tmp = this.read.readLine();
            return (tmp.equals("t")) ? true : false;
        } catch (IOException e) { return false; }

    }

    /**
     * Reads and returns queue objects from the server
     * @return a queue of the elements gathered from the server
     */
    public Queue<IgluMessage> ReadQueue() {
        Queue<IgluMessage> tmpQueue = new LinkedList<IgluMessage>(); // queue object to build

        while(CheckQueue()) {
            try {
                this.write.writeUTF("QR\n");
                String rawData = this.read.readLine();
                IgluMessage newMsg = new IgluMessage(rawData);
                tmpQueue.add(newMsg); // add next item to queue
            } catch (IOException e) { return tmpQueue; }
        }

        return tmpQueue;
    }
}

File: comm/interfaces_src/ServerCommunication.py
import socket
import queue
import time
import IgluMessage

# ServerCommunication class used to access interprocess communication server
# Programmer: Christian Wagner
# Date Created: 3/11/2020
# Date Modified: 11/02/2020
# Version: 2.0

class ServerCommunication:
    # Overloaded constructor
    # host -> the name of the host
    # port -> the port number being served on
    # pid -> the process identification number
    def __init__(self,host='localhost',port=8189,pid=0):
        self.__DEBUG_FLAG = True # debug messages flag
        self.__RETRY_CONNECTION_COUNT = 4 # number of times to retry before giving up
        self.__RETRY_CONNECTION_WAIT = 1 # time to wait, in seconds, between retries
        super().__init__() # call to super constructor
        self.__host = host # name of host
        self.__port = port # port number
        self.__pid = pid # process identifcation number
        self.__coordination_port = 18000 # port for communication coordination
        self.__s = 0 # setup socket variable
        self.__connected = False # default connection status

    # getter for process identification number variable
    @property
    def PID(self):
        return self.__pid

    # Connect to server using given host and port parameters
    def connect(self):
        self.__s = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # initialize socket object
        self.__s.connect((self.__host, self.__coordination_port)) # establish coordination socket connection
        
        try:
            tmp = 'test\n'
            self.__s.sendall(tmp.encode('utf-8')) # try sending on connection
            self.__connected = True
        except Exception as e:
            # connection failed, try again
            self.__connected = False

        if(not self.__connected):
            if(self.__DEBUG_FLAG): print('Connection on coordination port failed!')

            self.__s = 0 # clear variable

            # try to make connection again
            for x in range(0, self.__RETRY_CONNECTION_COUNT):
                if(self.__DEBUG_FLAG): print('Retrying connection on coordination port')

                self.__s = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # initialize socket object
                self.__s.connect((self.__host, self.__coordination_port)) # establish coordination socket connection

                time.sleep(self.__RETRY_CONNECTION_WAIT) # wait

                try:
                    tmp = 'test\n'
                    self.___s.sendall(tmp.encode('utf-8')) # try sending on connection
                    self.__connected = True # connecion was made
                except:
                    self.__connected = False # connection failed
                
                if self.__connected:
                    break

                # clear and restart
                self.__s = 0
            
            if not self.__connected:
                return False
                
        if(self.__DEBUG_FLAG): print('Connected to coordination port!')

        tmp = str(self.__port) + '\n'
        self.__s.sendall(tmp.encode('utf-8')) # communicate desired port
        check = self.__s.recv(1024) # get server status
        if check == b'a\n':
            self.disconnect()
            self.__s = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # initialize socket object
            self.__s.connect((self.__host, self.__port)) # establish socket connection
            ID = 'PID=' + str(self.__pid) + '\n' # create PID query
            self.__s.sendall(ID.encode('utf-8')) # send PID information
            return True
        else:
            if self.__DEBUG_FLAG: print('Coordination failed.')
            if self.__DEBUG_FLAG: print('Selected port is being used.')
            
            self.disconnect()
            return False

    # Disconnect from the server
    def disconnect(self):
        try:
            self.__s.sendall(b"exit\n") # signal exit
            self.__s.close() # close socket connection
            self.__s = 0 # clear variable
            return True
        except:
            return False

    # Open connection to the server
    def open(self):
        if self.__s != 0:
            try:
                self.__s.sendall(b"test\n") # signal exit
                return True # socket is already open
            except:
                return self.connect()
        else:
            return self.connect()

    # Close connection to the server
    def close(self):
        if self.__s == 0:
            return True # socket is already closed
        else:
            return self.disconnect()

    # Send data to the server for the queue
    # msg -> the data to send
    def WriteQueue(self, msg='NULL'):
        try:
            self.__s.sendall(b'QW\n')
            tmp = msg.get_tx_data() + '\n'
            self.__s.sendall(tmp.encode('utf-8')) # send data from IgluMessage object
            return True
        except:
            return False # socket is not connected

    # Check for items in the queue
    # return -> returns true if items are in the queue, false otherwise
    def CheckQueue(self):
        # method to check queue status
        self.__s.sendall(b'QC\n')
        check = self.__s.recv(1024) # get true/false flag
        if check == b't\n':
            return True # elemenets found
        else:
            return False # queue is empty

    # Reads and returns queue objects from the server
    # returns -> a queue of the elements gathered from the server
    def ReadQueue(self): # read items in queue
        tmpQueue = queue.Queue()
        while(self.CheckQueue()):
            try:
                self.__s.sendall(b'QR\n')
                data = self.__s.recv(1024) # get data
                data = str(data.decode('utf-8'))
                newMsg = IgluMessage.IgluMessage(data)
                tmpQueue.put(newMsg) # add fetched data to queue
            except:
                return tmpQueue
        return tmpQueue # return queue